relentless/optimize/method.py

```diff
@@ -177,41 +203,80 @@ def __init__(self, abs_tol, max_iter):
         self.abs_tol = abs_tol
         self.max_iter = max_iter
 
-    def find(self, objective, start, end):
+    def find(self, objective, start, end, scale=1.0):
         """Apply the line search algorithm to take the optimal step.
 
         Note that the objective function is kept at its initial state, and the
         function evaluted after taking the optimal step is returned separately.
 
         Parameters
         ----------
-        objective : :class:`~relentless.optimize.ObjectiveFunction`
+        objective : :class:`~relentless.optimize.objective.ObjectiveFunction`
             The objective function for which the line search is applied.
-        start : :class:`~relentless.optimize.ObjectiveFunctionResult`
+        start : :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
             The objective function evaluated at the start of the search interval.
-        end : :class:`~relentless.optimize.ObjectiveFunctionResult`
+        end : :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
             The objective function evaluated at the end of the search interval.
+        scale : float or dict
+            A scalar scaling parameter or scaling parameters (:math:`\mathbf{X}`)
+            keyed on one or more `~relentless.optimize.objective.ObjectiveFunction`
+            design variables (defaults to ``1.0``, so that the variables are unscaled).
 
         Raises
         ------
+        ValueError
+            If the scaling parameters are not positive.
+        ValueError
+            If the start and the end of the search interval are identical.
+        KeyError
+            If the scaling parameters are defined on a variable which is not in
+            the objective function.
         ValueError
             If the defined search interval is not a descent direction.
 
         Returns
         -------
-        :class:`~relentless.optimize.ObjectiveFunctionResult`
-            The objective function evaluated at the new, 'optimal' location.
+        :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
+            The objective function evaluated at the new, "optimal" location.
 
         """
+        # check if valid scaling
+        try:
+            sc = dict(scale)
+            err = any([s <= 0 for s in scale.values()])
+        except TypeError:
+            sc = scale
+            err = scale <= 0
+        if err:
+            raise ValueError('The scaling parameters must be positive.')
+
         ovars = {x: x.value for x in objective.design_variables()}
 
         # compute normalized search direction
         d = end.design_variables - start.design_variables
         max_step = d.norm()
+        if max_step == 0:
+            raise ValueError('The start and end of the search interval must be different.')
         d /= max_step
 
+        # compute scaled gradients
+        grad_y_start = start.gradient
```

Comment:
I find this notation confusing. This is actually `grad_x_start`. I would rather that you left these variables in their original form, and then multiply by the scale factors where they are needed. It will make the intent a lot cleaner, and you can do any error checking once above.

Reply:
Now that we are not scaling, this can just be `grad_start`

Reply:
Even better, you can just use `start.gradient` without assigning a new variable :-)